implementations/rust/ockam/ockam_command/src/upgrade.rs

```diff
@@ -1,51 +1,149 @@
+            "This will upgrade ockam to the latest version. Are you sure?"
+        ))? {
+            ConfirmResult::Yes => {}
+            ConfirmResult::No => {
+                return Ok(());
+            }
+            ConfirmResult::NonTTY => {
+                return Err(miette!("Use --yes to confirm"));
             }
         }
     }
+    opts.terminal.write_line(fmt_info!(
+        "Upgrading ockam from {} to {}",
+        current_version,
+        latest_version
+    ))?;
+
+    upgrade_ockam(&opts)?;
+    opts.terminal
+        .stdout()
+        .plain(fmt_ok!("Ockam upgraded to version {}", latest_version))
+        .write_line()?;
+    Ok(())
+}
+
+fn stop_all_running_nodes(opts: &CommandGlobalOpts) -> Result<Vec<String>> {
+    opts.terminal
+        .write_line(fmt_info!("Stopping all running nodes"))?;
+    let nodes_states = opts.state.nodes.list()?;
+    let mut stopped_nodes = Vec::new();
+    for node_state in nodes_states.iter() {
+        if node_state.is_running() {
+            node_state.kill_process(false)?;
+            opts.terminal
+                .write_line(fmt_ok!("Stopped node {}", node_state.name()))?;
+            stopped_nodes.push(node_state.name().to_string());
+        }
+    }
+    Ok(stopped_nodes)
+}
+
+fn restart_nodes(stopped_nodes_names: &[String], opts: &CommandGlobalOpts) -> miette::Result<()> {
+    opts.terminal
+        .write_line(fmt_info!("Restarting all stopped nodes"))?;
+    for node_name in stopped_nodes_names.iter() {
+        let node_state = opts.state.nodes.get(node_name)?;
+        let node_setup = node_state.config().setup();
+        spawn_node(
+            &opts,
+            &node_name,                                    // The selected node name
+            &node_setup.api_transport()?.addr.to_string(), // The selected node api address
+            None,                                          // No project information available
+            None,                                          // No trusted identities
+            None,                                          // "
+            None,                                          // "
+            None,                                          // Launch config
+            None,                                          // Authority Identity
+            None,                                          // Credential
+            None,                                          // Trust Context
+            None,                                          // Project Name
+            node_setup.disable_file_logging,
+        )?;
+        opts.terminal
+            .write_line(fmt_ok!("Restarted node {}", node_state.name()))?;
+    }
+    Ok(())
 }
 
-fn upgrade_check_is_disabled() -> bool {
-    get_env_with_default("OCKAM_DISABLE_UPGRADE_CHECK", false).unwrap_or(false)
+fn upgrade_ockam(opts: &CommandGlobalOpts) -> miette::Result<()> {
+    let stopped_nodes_names = stop_all_running_nodes(opts)?;
```

Comment:
Wouldn't it be safer to move this at least after `let installer = get_installer(opts)?;`? Or, even better, right before restarting the nodes. I'm not sure what will happen to running processes after doing the upgrade (are they killed after removing the current binary version?).

Reply:
So, if you run the upgrade without stopping the nodes, everything will still be running, so we could only do the restart cycle after the upgrade has run successfully just to ensure the nodes get upgraded to the latest version.

As an aside to this, with all the changes made with the nodes logic since the `0.90.0` release, the nodes will no longer be restarted correctly on this branch. I'm pretty sure this is just due to the nodes trying to be restarted with the `0.90.0` version after you run the upgrade? I created a new branch https://github.com/BradLewis/ockam/tree/bradlewis/upgrade-test just to test this and everything is restarted as expected.

Reply:
Sure.
1. Create a node using `ockam node create n` using any version of `ockam`.
2. Change the version in the `Cargo.toml` file for `ockam_command` to something below `0.90.0`.
3. Run `cargo run upgrade --yes` to "upgrade" the ockam binary to the latest version. This should just replace the one in `target/debug/ockam` with the latest one from github.
4. Everything appears to run smoothly, however if you run `cargo run node list`, you'll see that the node is stopped.

I'm pretty sure this is because it's trying to restart the nodes using the `0.90.0` release binary, but the "new" nodes are not compatible with that version.

If you run `ockam node list` you'll see that it thinks there are no nodes on the system 
```shell
❯ ockam node list
       ┌───────────────────┐
       │       Nodes       │
       └───────────────────┘

     ! No nodes found on this system.

❯ cargo run node list
    Finished dev [unoptimized + debuginfo] target(s) in 0.20s
     Running `/Users/bradlewis/repos/ockam/target/debug/ockam node list`
       ┌───────────────────┐
       │       Nodes       │
       └───────────────────┘

     │ Node n (default) UP
     │ Process id 21821

     │ Node m  UP
     │ Process id 22503

```

You can also easily reproduce this behaviour by first creating a node with release `0.90.0`, then a new one with `cargo run` and then listing the nodes

```shell
❯ ockam node create n
       Creating Node n...

     ✔︎ Node n created successfully

       To see more details on this node, run:
       ockam node show

❯ ockam node list
       ┌───────────────────┐
       │       Nodes       │
       └───────────────────┘

     │ Node n (default) UP
     │ Process id 26852


❯ cargo run node create m
    Finished dev [unoptimized + debuginfo] target(s) in 0.20s
     Running `/Users/bradlewis/repos/ockam/target/debug/ockam node create m`
       Creating Node m...

     ✔︎ Node m created successfully

       To see more details on this node, run:
       ockam node show

❯ cargo run node list
    Finished dev [unoptimized + debuginfo] target(s) in 0.20s
     Running `/Users/bradlewis/repos/ockam/target/debug/ockam node list`
       ┌───────────────────┐
       │       Nodes       │
       └───────────────────┘

     │ Node n (default) UP
     │ Process id 26852

     │ Node m  UP
     │ Process id 27086

❯ ockam node list
       ┌───────────────────┐
       │       Nodes       │
       └───────────────────┘

     ! No nodes found on this system.

```